File: Instructors Area/Instructor-Materials/Spark-Essentials/Labs/Optional/DAG-Stages-Tasks/dag-stages-arch-demo.scala
// Databricks notebook source exported at Sun, 31 Jan 2016 00:29:32 UTC
//Start this demo on a Spark cluster with at least 1 Worker

// COMMAND ----------

// Where will the #5 print here? Which JVM's log?
// Answer: Driver JVM
val y = 4 + 1
println(y)

// COMMAND ----------

// First, create a Scala array
val wordList = Array("cat", "cat", "fish", "dog", "fish")

// COMMAND ----------

// Now, we'll create an RDD from the Scala array, requesting 1 partition
val wordsRDD = sc.parallelize(wordList, 1)

// COMMAND ----------

// Show # of tasks in UI + expand below
wordsRDD.count()

// COMMAND ----------

wordsRDD.partitions.length

// COMMAND ----------

// Also run this a few times and show that on the Executors UI page, the # of tasks just goes up on 1 Executor (not both)
wordsRDD.collect()

// COMMAND ----------

//Why won't this print anything?
// Answer: because there's no action
wordsRDD.map(c => println(c))

// COMMAND ----------

// Where will this print occur? 
// A: To the logs for the Executors (show students via Spark UI's Executors page)
wordsRDD.map(c => println(c)).collect()

// COMMAND ----------

val wordsRDDWith2Partitions = sc.parallelize(wordList, 2)

// COMMAND ----------

wordsRDDWith2Partitions.collect()

// COMMAND ----------

// How do we check what is in each of the two parititions?
wordsRDDWith2Partitions.map(c => println(c)).collect()

// COMMAND ----------

for (p <- wordsRDDWith2Partitions.glom.collect())
  println(p.toSeq)

// COMMAND ----------

sc

// COMMAND ----------

sc.version

// COMMAND ----------

sc.appName

// COMMAND ----------

sc.defaultParallelism

// COMMAND ----------

// Notice that this has 1 task per executor core
val testRDD = sc.parallelize(0 to 100).collect()

// COMMAND ----------

display(dbutils.fs.ls("dbfs:///mnt/training/data/tom_sawyer"))

// COMMAND ----------

// Show how the IO aligns in UI
sc.textFile("dbfs:///mnt/training/data/tom_sawyer", 2).collect()

// COMMAND ----------

// What will be the type of ebook below?
val ebook = sc.textFile("dbfs:///mnt/training/data/tom_sawyer").collect()

// COMMAND ----------

// Why doesnt this work?
ebook.count()

// COMMAND ----------

ebook.getClass

// COMMAND ----------

val ebookRDD = sc.textFile("dbfs:///mnt/training/data/tom_sawyer", 2)

// COMMAND ----------

ebookRDD.count()

// COMMAND ----------

ebookRDD.partitions.length

// COMMAND ----------

ebookRDD.getClass

// COMMAND ----------

ebookRDD

// COMMAND ----------

// It simply counts the number of items in each partition.
for ((p, i) <- ebookRDD.glom.collect().zipWithIndex)
  println(s"$i: ${p.length} item(s)")

// COMMAND ----------

// Show how this is 2 operations in the UI, STILL 2 tasks! Pipelining...
sc.textFile("dbfs:///mnt/training/data/tom_sawyer", 2).flatMap(line => line.split(" ")).collect()

// COMMAND ----------

sc.textFile("dbfs:///mnt/training/data/tom_sawyer", 2).flatMap(line => line.split(" ")).map(word => (word, 1)).collect()


// COMMAND ----------

sc.textFile("dbfs:///mnt/training/data/tom_sawyer", 2).flatMap(line => line.split(" ")).map(word => (word, 1)).reduceByKey((x, y) => x + y).collect()


// COMMAND ----------

// Notice stage 2 has 4 tasks now
sc.textFile("dbfs:///mnt/training/data/tom_sawyer", 2).flatMap { line => line.split(" ") }.map { word => (word, 1) }.reduceByKey((x, y) => x + y, 4).collect()


// COMMAND ----------

//why doesnt this show up?
val cachedBookRDD = sc.textFile("dbfs:///mnt/training/data/tom_sawyer", 2).cache()

// COMMAND ----------

cachedBookRDD.setName("My Cached Book") 

// COMMAND ----------

// Run this twice to show speed increase
// Show size in memory of RDD
cachedBookRDD.count()

// COMMAND ----------

cachedBookRDD.unpersist()

// COMMAND ----------

//why doesnt this show up?
import org.apache.spark.storage.StorageLevel._

val cachedBookSerializedRDD = sc.textFile("dbfs:///mnt/training/data/tom_sawyer", 2).persist(MEMORY_ONLY_SER)

// COMMAND ----------

cachedBookSerializedRDD.setName("My Serialized Cached Book")

// COMMAND ----------

cachedBookSerializedRDD.count()

// COMMAND ----------

// Show green cache dot in UI
sc.textFile("dbfs:///mnt/training/data/tom_sawyer", 2).cache().flatMap(line => line.split(" ")).map(word => (word, 1)).reduceByKey((x, y) => x + y, 4).collect()


// COMMAND ----------

val tempRDD = sc.textFile("dbfs:///mnt/training/data/tom_sawyer", 2).repartition(4)

// COMMAND ----------

tempRDD.partitions.length

// COMMAND ----------

// It simply counts the number of items in each partition.
for ((p, i) <- tempRDD.glom.collect().zipWithIndex)
  println(s"$i: ${p.length} item(s)")

// COMMAND ----------

sc.textFile("dbfs:///mnt/training/data/tom_sawyer", 2).repartition(4).cache().flatMap(line => line.split(" ")).map(word => (word, 1)).reduceByKey((x, y) => x + y, 4).collect()

// COMMAND ----------

val xRDD = sc.textFile("dbfs:///mnt/training/data/tom_sawyer").repartition(4).cache()

xRDD.flatMap(line => line.split(" ")).map(word => (word, 1)).reduceByKey((x, y) => x + y, 4).collect()

// COMMAND ----------

// Show how this lets you skip the earlier Stage in UI
// Change the word from Hadoop to another word like Tom
xRDD.filter(line => line.contains("Tom")).count()

// COMMAND ----------

val myRDD = xRDD.flatMap(line => line.split(" ")).map(word => (word, 1)).reduceByKey((x, y) => x + y, 4).collect()

// COMMAND ----------

val myRDD = xRDD.flatMap(line => line.split(" ")).map(word => (word, 1)).reduceByKey((x, y) => x + y, 4).sample(false, 0.01, 42).collect()

File: Spark-DataScience-TD1/Spark-Essentials/Labs/Optional/DAG-Stages-Tasks/dag-stages-arch-demo.py
# Databricks notebook source exported at Thu, 14 Jan 2016 00:08:58 UTC
# First, create a Python list
wordList = ["doug", "jon", "sameer", "eliano", "richard"]

# COMMAND ----------

# Now, we'll create an RDD from the Python list with 4 partitions
wordsRDD = sc.parallelize(wordList, 4)

# COMMAND ----------

# Why won't this print anything?
def println(value):
   print value
    
wordsRDD.map(println)

# COMMAND ----------

# This will print something... but where?
wordsRDD.map(println).collect()
# Let's look in (executor) logs

# COMMAND ----------

# Now, we'll create an RDD from the array, requesting 6 partitions
wordsRDD = sc.parallelize(wordList, 6)

# COMMAND ----------

wordsRDD.collect()

# COMMAND ----------

sc

# COMMAND ----------

sc.version

# COMMAND ----------

sc.appName

# COMMAND ----------

sc.defaultParallelism

# COMMAND ----------

datapath = "dbfs:/mnt/training/data/tom_sawyer" # https://www.gutenberg.org/ebooks/74

# COMMAND ----------

sc.textFile(datapath, 2).collect()

# COMMAND ----------

# Note, collect does not return an RDD (it returns an Array)
ebook = sc.textFile(datapath, 2).collect()

# COMMAND ----------

# Why doesnt this work?
ebook.count()

# COMMAND ----------

print 'type of ebook: {0}'.format(type(ebook))

# COMMAND ----------

# Why doesnt this show up in the UI
ebookRDD = sc.textFile(datapath, 2)

# COMMAND ----------

ebookRDD.count()

# COMMAND ----------

ebookRDD.getNumPartitions()

# COMMAND ----------

print 'type of ebookRDD: {0}'.format(type(ebookRDD))

# COMMAND ----------

ebookRDD

# COMMAND ----------

# Let's use RDD.glom to look at the distribution of data in our partitions

# What does glom do? Replaces each partition in the RDD with a single container that contains all of the partition's previous items
# http://spark.apache.org/docs/latest/api/python/pyspark.html#pyspark.RDD.glom

glomDemo = sc.parallelize([1,2,3,4,5,6,7,8,9], 3)
glomDemo.glom().collect()

# How does it work? This is a good one to guess about, and then check the source

# COMMAND ----------

ebookRDD.glom().collect()

# COMMAND ----------

# Count the number of items in each partition.
for (p, i) in ebookRDD.glom().zipWithIndex().collect():
  print '%d: %d items(s)' % (i, len(p))

# COMMAND ----------

sc.textFile(datapath, 2).flatMap(lambda x: x.split(' ')).collect()

# COMMAND ----------

sc.textFile(datapath, 2).flatMap(lambda x: x.split(' ')).map(lambda s: (s, 1)).collect()

# COMMAND ----------

sc.textFile(datapath, 2).flatMap(lambda x: x.split(' ')).map(lambda s: (s, 1)).reduceByKey(lambda x, y: x + y).collect()

# COMMAND ----------

# why doesnt this show up?
cachedBookRDD = sc.textFile(datapath, 2).cache()

# COMMAND ----------

cachedBookRDD.setName("My Cached Book") 

# COMMAND ----------

cachedBookRDD.count()

# COMMAND ----------

cachedBookRDD.unpersist()

# COMMAND ----------

from pyspark.storagelevel import StorageLevel
cachedBookSerializedRDD = sc.textFile(datapath, 2).persist(StorageLevel.MEMORY_ONLY_SER) #same thing as above, since we're using Python

# COMMAND ----------

cachedBookSerializedRDD.setName("My Serialized Cached Book")

# COMMAND ----------

cachedBookSerializedRDD.count()

# COMMAND ----------

sc.textFile(datapath, 2).flatMap(lambda x: x.split(' ')).map(lambda s: (s, 1)).reduceByKey(lambda x, y: x + y).collect()

# COMMAND ----------

tempRDD = sc.textFile(datapath, 2).repartition(4)

# COMMAND ----------

tempRDD.getNumPartitions()

# COMMAND ----------

# It simply counts the number of items in each partition.
for (p, i) in tempRDD.glom().zipWithIndex().collect():
  print '%d: %d items(s)' % (i, len(p))

# COMMAND ----------

sc.textFile(datapath, 2).repartition(12).cache().flatMap(lambda x: x.split(' ')).map(lambda s: (s, 1)).reduceByKey(lambda x, y: x + y).collect()

# COMMAND ----------

sc.textFile(datapath, 2).repartition(4).cache().flatMap(lambda x: x.split(' ')).map(lambda s: (s, 1)).reduceByKey(lambda x, y: x + y).collect()

# COMMAND ----------

myRDD = sc.textFile(datapath, 2).repartition(12).cache().flatMap(lambda x: x.split(' ')).map(lambda s: (s, 1)).reduceByKey(lambda x, y: x + y, numPartitions=8)
myRDD.collect()

# COMMAND ----------

myRDD.getNumPartitions()

# COMMAND ----------

myRDD.count() 
# this job uses data that is downstream from a shuffle we recently did -- Spark should use those shuffle output files to take a shortcut
# let's look in the UI for skipped stages
